### PR TITLE
[EDA] Settings Saving Functionality

### PR DESCRIPTION
**Pull request by m-kinkade**
# Critical Changes

# Changes

# Issues Closed

# New Metadata
### Custom Labels
stgErrorInsufficientAccessSaving
stgErrorInvalidSettings

### Apex Classes
HierarchySettingsChangesController
HierarchySettingsChangesController_TEST
HierarchySettingsService
HierarchySettingsService_TEST
HierarchySettingsMapper
HierarchySettingsMapper_TEST
HierarchySettingsChangesModel
HierarchySettingsChangesModel_TEST
HierarchySettingsModel
HierarchySettingsModel_TEST

## Unpackaged Metadata

# Deleted Metadata

# Testing Notes

### CODE REVIEW

File: force-app/main/default/classes/HierarchySettingsChangesModel.cls [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+
+    /**********************************************************************************************************************
+    * @description No argument constructor
+    * @return A HierarchySettingsChangesModel with empty maps
+    **********************************************************************************************************************/
+    public HierarchySettingsChangesModel() {
+        this.settingsSingleValueBySettingsName = new Map<String, String>();
+        this.settingsListSettingsName = new Map<String, List<String>>();
+
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/HierarchySettingsChangesModel.cls:46]: This extra line feels odd here.

File: force-app/main/default/classes/HierarchySettingsChangesModel.cls [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+/**
+* @author Salesforce.org
+* @date 2021
+* @group 
+* @group-content ../../ApexDocContent/EdaSettings.htm
+* @description Model to abstract data for changes made to Hierarchy Settings.
+*/
+public with sharing class HierarchySettingsChangesModel {
+    public Map<String, String> settingsSingleValueBySettingsName { get; set; }
+    public Map<String, List<String>> settingsListSettingsName { get; set; }
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/HierarchySettingsChangesModel.cls:36]: These need to be aura enabled

File: force-app/main/default/classes/HierarchySettingsChangesModel.cls [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2021
+* @group 
+* @group-content ../../ApexDocContent/EdaSettings.htm
+* @description Model to abstract data for changes made to Hierarchy Settings.
+*/
+public with sharing class HierarchySettingsChangesModel {
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/HierarchySettingsChangesModel.cls:34]: Now that we are pulling it down, we probably should have this be a view model instead.
```suggestion
public with sharing class HierarchySettingsChangesVModel {
```

File: force-app/main/default/classes/HierarchySettingsService.cls [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    @TestVisible
+    private HierarchySettingsMapper locateHierarchySettingsMapper() {
+        return HierarchySettingsMapper.getInstance();
+    }
+
+    /************************************************************************************************************
+    * @description Retrieve an instance of UserPermissionService
+    * @return An instance of the UserPermissionService class
+    ************************************************************************************************************/ 
+    // @TestVisible
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/HierarchySettingsService.cls:81]: This is still commented out. Are we still waiting on that service? We can merge in the other branch and define this as a dependency for it.

File: force-app/main/default/classes/HierarchySettingsService_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            HierarchySettingsMapper.getInstance(), 
+            hierarchySettingsMapperInstance, 
+            'Should return instance of HierarchySettingsMapper.');
+    }
+
+    /**************************************************************************************************************************
+    * @description Test method to verify that the locateUserPermissionService method returns a valid instance of 
+    * UserPermissionService.
+    ***************************************************************************************************************************/
+    // @isTest
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/HierarchySettingsService_TEST.cls:113]: Likewise here for the commenting

File: force-app/main/default/classes/HierarchySettingsMapper.cls [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            UTIL_Debug.debug('****SchemaDescribeException encountered: ' + e.getMessage());
+            throw new InvalidSettingsException(Label.stgErrorInvalidSettings);
+        } catch (System.TypeException e) {
+            UTIL_Debug.debug('****TypeException encountered: ' + e.getMessage());
+            throw new InvalidSettingsException(Label.stgErrorInvalidSettings);
+        } catch (Exception e) {
+            UTIL_Debug.debug('****Unexpected Exception encountered: ' + e.getMessage());
+            throw new InvalidSettingsException(Label.stgErrorInvalidSettings);
+        }
+    }
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/HierarchySettingsMapper.cls:140]: This method should be moved into the service.

File: force-app/main/default/classes/HierarchySettingsMapper.cls [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    /******************************************************************************************************
+    * @description Converts a List of Strings to a semicolon-delimited string.
+    * @param  collectionTypeHierarchySettingsUpdateList List<String> to convert
+    * @return A semicolon-delimited String representing each String in the specified list.
+    ******************************************************************************************************/
+    @TestVisible
+    private String denormalizeListSettingChange(List<String> collectionTypeHierarchySettingsUpdateList) {
+        return String.join(collectionTypeHierarchySettingsUpdateList, ';');
+    }
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/HierarchySettingsMapper.cls:150]: This method should be moved into the service.

**Review by MatthewBlanski — CHANGES_REQUESTED** [review]
A few changes requested.

**Comment by m-kinkade** [comment]
This branch has dependencies on the UserPermissionsService.  Merge this branch AFTER https://github.com/SalesforceFoundation/EDA/pull/1581

File: force-app/main/default/lwc/settingsSaveCanvas/settingsSaveCanvas.js [hunk 8]
```diff
@@ -1,47 +1,110 @@
...
+                }
+            });
+    }
+
+    clearHierarchySettingsChanges() {
+        this.hierarchySettingsChanges.settingsSingleValueBySettingsName = {};
+        this.hierarchySettingsChanges.settingsListSettingsName = {};
+    }
+
+    completeSave() {
```
[the remark this answers is not in the document]
Comment (MatthewBlanski) [reply force-app/main/default/lwc/settingsSaveCanvas/settingsSaveCanvas.js:83]: Let's excise it here @m-kinkade

**Review by MatthewBlanski — APPROVED** [review]
This looks like it's working! I'm so excited.

**Comment by spinkelman** [comment]
Limited QA passes - Checked over code and labels, but further testing needed once functioning fields are added.

**Pull request merged**